segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -493,35 +588,71 @@ private SystemSnapshotRecord applySystemSnapshotRecord(MetadataTransaction txn,
         log.debug("SystemJournal[{}] Done applying snapshots.", containerId);
 
         // Validate
-        validateSystemSnapshot(systemSnapshot);
-        validateSystemSnapshotExistsInTxn(txn, systemSnapshot);
-        return systemSnapshot;
+        return checkInvariants(systemSnapshot)
+                .thenComposeAsync(v ->
+                        validateSystemSnapshotExistsInTxn(txn, systemSnapshot), executor)
+                .thenApplyAsync(v -> systemSnapshot, executor);
     }
 
-    private void validateSystemSnapshotExistsInTxn(MetadataTransaction txn, SystemSnapshotRecord systemSnapshot) throws Exception {
-        if (null != systemSnapshot) {
-            for ( val segmentSnapshot : systemSnapshot.getSegmentSnapshotRecords()) {
-                Preconditions.checkState(null != txn.get(segmentSnapshot.segmentMetadata.getKey()).get());
-                for ( val metadata : segmentSnapshot.getChunkMetadataCollection()) {
-                    Preconditions.checkState(null != txn.get(metadata.getKey()).get());
-                }
-                validateSegmentRecordInTxn(txn, segmentSnapshot.segmentMetadata.getKey());
-            }
+    /**
+     * Validate that all information in given snapshot exists in transaction.
+     */
+    private CompletableFuture<Void> validateSystemSnapshotExistsInTxn(MetadataTransaction txn, SystemSnapshotRecord systemSnapshot) {
+        if (null == systemSnapshot) {
+            return CompletableFuture.completedFuture(null);
+        } else {
+            // For each segment in snapshot
+            return Futures.loop(
+                    systemSnapshot.getSegmentSnapshotRecords(),
+                    segmentSnapshot ->
+                        txn.get(segmentSnapshot.segmentMetadata.getKey())
+                            .thenApplyAsync(metadata -> {
+                                Preconditions.checkState(null != metadata);
+                                return metadata;
+                            }, executor)
+                            .thenComposeAsync(v ->
+                                // For each chunk in the segment
+                                Futures.loop(
```

Comment:
I suggest you move this inner for loop into its own method to improve readability.

Reply:
done